language/tools/move-build/src/source_package/manifest_parser.rs

```diff
@@ -0,0 +1,447 @@
+                        }
+                        let (rename, assignment) = rename_subst.into_iter().next().unwrap();
+                        // Try and parse to an address, otherwise it's an error
+                        let rename_from = Identifier::new(rename.as_str()).context("Invalid address name")?;
+                        let assignment = assignment.as_str().ok_or_else(|| format_err!("Invalid assignment in substitution {}. Address must be a string", assignment))?;
+                        let assignment = AccountAddress::from_hex_literal(assignment).context("Invalid address")?;
+                        subst.insert(addr_ident, PM::SubstOrRename::RenameFromAndAssign{
+                            rename_from,
+                            assignment,
+                        });
+                    }
+                    TV::String(addr_or_name) => {
+                        if let Ok(addr) = AccountAddress::from_hex_literal(&addr_or_name) {
+                            subst.insert(addr_ident, PM::SubstOrRename::Assign(addr));
+                        } else {
+                            let rename_from = Identifier::new(addr_or_name.as_str()).context("Invalid address name")?;
+                            subst.insert(addr_ident, PM::SubstOrRename::RenameFrom(rename_from));
+                        }
+                    }
+                    x => bail!(
+                        "Ill-formed dependency substitution {}. Expected a table or string, but encountered a {}",
+                        x,
+                        x.type_str()
+                    ),
+                }
+            }
+            Ok(subst)
+        }
+        x => bail!(
+            "Ill-formed dependency substitution {}. Expected a table, but encountered a {}",
+            x,
+            x.type_str()
+        ),
+    }
+}
+
+fn parse_version(tval: TV) -> Result<PM::Version> {
+    let version_str = tval.as_str().unwrap();
+    let version_parts = version_str.split('.').collect::<Vec<_>>();
+    if version_parts.len() != 3 {
+        bail!(
+            "Version is malformed. Versions must be of the form <u64>.<u64>.<u64>, but found '{}'",
+            version_str
+        );
+    }
+
+    Ok((
+        version_parts[0]
+            .parse::<u64>()
+            .context("Invalid major version")?,
+        version_parts[1]
+            .parse::<u64>()
+            .context("Invalid minor version")?,
+        version_parts[2]
+            .parse::<u64>()
+            .context("Invalid bugfix version")?,
+    ))
+}
+
+// check that only recognized names are provided at the top-level
+fn check_for_unknown_field_names(
+    table: &toml::map::Map<String, TV>,
+    known_names: &[&str],
+) -> Result<()> {
+    let mut unknown_names = BTreeSet::new();
+    for key in table.keys() {
+        if !known_names.contains(&key.as_str()) {
+            unknown_names.insert(key.to_string());
+        }
+    }
+
+    if !unknown_names.is_empty() {
+        bail!(
```

Comment:
As you mentioned in the PR description, it might be best to make this a non-fatal warning.